File: src/p0998/java/solution.java
class Solution {
    public TreeNode insertIntoMaxTree(TreeNode root, int val) {
        TreeNode sentinel = new TreeNode(0); sentinel.right = root;
        TreeNode prev = sentinel, curr = root;
        while (curr != null && curr.val > val) {
            prev = curr; curr = curr.right;
        }
        prev.right = new TreeNode(val);
        prev.right.left = curr;
        return sentinel.right;
    }
}

File: src/p0998/golang/solution.go
package main

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

func insertIntoMaxTree(root *TreeNode, val int) *TreeNode {
	sentinel := &TreeNode{Right: root}
	prev, curr := sentinel, root
	for curr != nil && curr.Val > val {
		prev, curr = curr, curr.Right
	}
	prev.Right = &TreeNode{Val: val, Left: curr}
	return sentinel.Right
}
